packages/core/src/browser/shell.ts

```diff
@@ -56,21 +66,83 @@ export interface DockLayoutData extends DockPanel.ILayoutConfig {
     activeWidgets?: Widget[]
 }
 
+@injectable()
+export class DockPanelTabBarRenderer implements TabBar.IRenderer<any> {
+    readonly closeIconSelector = TabBar.defaultRenderer.closeIconSelector;
+
+    constructor( @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer) { }
+    /**
+     * Render the virtual element for a tab.
+     *
+
+     * @param data - The data to use for rendering the tab.
+     *
+     * @returns A virtual element representing the tab.
+     */
+    renderTab(data: TabBar.IRenderData<any>): VirtualElement {
+        const title = data.title.caption;
+        const key = TabBar.defaultRenderer.createTabKey(data);
+        const style = TabBar.defaultRenderer.createTabStyle(data);
+        const className = TabBar.defaultRenderer.createTabClass(data);
+        const dataset = TabBar.defaultRenderer.createTabDataset(data);
+        return (
+            h.li({
+                key, className, title, style, dataset,
+                oncontextmenu: event => this.handleContextMenuEvent(event)
+            },
+                TabBar.defaultRenderer.renderIcon(data),
+                TabBar.defaultRenderer.renderLabel(data),
+                TabBar.defaultRenderer.renderCloseIcon(data)
+            )
+        );
+    }
+
+    handleContextMenuEvent(event: MouseEvent) {
+        event.stopPropagation();
+        event.preventDefault();
+        this.contextMenuRenderer.render(MAINAREA_TABBAR_CONTEXT_MENU, event);
+    }
+}
+
+/* Note this class must be below DockPanelTabBarRender othewise inversify thows an error:
```

Comment:
In @inject you don't reference a class as a type but as a service identifier (a symbol) which should be declared before it can be used.

I don't think such comments are helpful.

Reply:
But in javascript/typescript usually symbols are "top promoted" right like meaning as long as they are defined in the same file I don't have to declare something before using it no ?

In this case should I not be able to exchange the order of these classes ?

Reply:
Only the declaration is hoisted to the top, the assignment is not.

Reply:
Haa k k so the value isn't. thx!